 <img align="left" width="75" height="75"  src="https://upload.wikimedia.org/wikipedia/en/c/c8/University_of_the_Punjab_logo.png"> 

<h1 align="center">Department of Data Science</h1>
<h1 align="center">Course: Tools and Techniques for Data Science</h1>

---
<h3><div align="right">Instructor: Muhammad Arif Butt, Ph.D.</div></h3>    

<h1 align="center">Lecture 6.26 (Bayes Theorem and Naïve Bayes' Classifiers)</h1>

<a href="https://colab.research.google.com/github/arifpucit/data-science/blob/master/Section-4-Mathematics-for-Data-Science/Lec-4.1(Descriptive-Statistics).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <img align="center" width="900" src="images/ml-topimg1.png"  >

## Learning agenda of this notebook
- Bayes' Theorem
- Proof of Bayes' Theorem
- Formulas of Bayes' Theorem
- Examples of Bayes' Theorem
- Overview of Naïve Bayes' Classifier
- Naïve Bayes' Classifier for Datasets with Discrete Input Features
    - Example 1: Single Input Feature
    - Example 2: Multiple Input Features
    - Example 3: Multiple Input Features
    - Example 4: Using Naïve Bayes' on Text Data
- Naïve Bayes' Classifier for Datasets with Continuous Input Features
    - Example 1: Single Input Feature
    - Example 2: Multiple Input Features
- Tasks to Do

# 1.  Bayes’ Theorem

- **Bayes' Theorem** (by Thomas Bayes), is a way of calculating a conditional probability without the joint probability.
- The theorem was introduced by Thomas Bayes, a statistician and philosopher and was first published in 1763.
- Bayes Theorem is the corner stone of Bayesian Learning methods that is a paradigm for constructing statistical models based on Bayes' Theorem

<font face = "Courier New" color=magenta size=6>$$     P(A \mid B) \hspace{0.5cm} = \hspace{0.5cm}  \frac{P(A) \times P(B \mid A)}{P(B)}, \hspace{0.5cm} P(B)\neq 0 \hspace{0.5cm} $$</font>

## a.  Proof of Bayes’ Theorem 

- To calculate the probability of event A to occur, given that event B has already occurred, we can use the following **Conditional Probability** formula in equation (i):


\begin{equation}
    P(A \mid B) = \frac{P(A \cap B)}{P(B)} \hspace{0.5cm} ------(i)
\end{equation}

- Similarly, the probability of event B, given that event A has already occurred is given by equation2:

\begin{equation}
    P(B \mid A) = \frac{P(B \cap A)}{P(A)} \hspace{0.9cm} OR \hspace{0.9cm}  P(B \mid A) = \frac{P(A \cap B)}{P(A)} \hspace{0.5cm}------(ii)
\end{equation}

- Multiplying both sides of equation $(i)$ by $P(B)$ gives us:

\begin{equation}
    P(A \mid B) * P(B) = P(A \cap B) \hspace{0.5cm} ------(iii)
\end{equation}


- Similarly, multiplying both sides of equation $(ii)$ by $P(A)$ gives us:

\begin{equation}
    P(B \mid A) * P(A) = P(A \cap B) \hspace{0.5cm} ------(iv)
\end{equation}

- Equating equations $(iii)$ and $(iv)$, we get:

\begin{equation}
P(A \mid B) * P(B) \hspace{0.5cm} = \hspace{0.5cm}  P(B \mid A) * P(A)  \hspace{0.5cm}
\end{equation}

$$     P(A \mid B) \hspace{0.5cm} = \hspace{0.5cm}  \frac{P(B \mid A) \times P(A)}{P(B)}, \hspace{0.5cm} P(B)\neq 0 \hspace{0.5cm} $$

## b. Formulas of Bayes' Theorem
<font face = "Courier New" color=magenta size=6>$$     P(A \mid B) \hspace{0.5cm} = \hspace{0.5cm}  \frac{P(A) \times P(B \mid A)}{P(B)}, \hspace{0.5cm} P(B)\neq 0 \hspace{0.5cm} $$</font>
<br><br>
<font face = "Courier New" color=blue size=6>$$     P(h \mid D) \hspace{0.5cm} = \hspace{0.5cm}  \frac{P(h) \times P(D \mid h)}{P(D)}, \hspace{0.5cm} P(D)\neq 0 \hspace{0.5cm} $$</font>
<br><br>
<font face = "Courier New" color=green size=6>$$     P(y=k \mid X) \hspace{0.5cm} = \hspace{0.5cm}  \frac{P(y=k) \times P(X \mid y=k)}{P(X)}, \hspace{0.5cm} P(X)\neq 0 \hspace{0.5cm} $$</font>


- Where:
    >- $P(y=k \mid X)$ is `Posterior Probability`: Probability of output label y given the input features X (also known as conditional probability).
    >- $P(X|y=k)$ is `Liklihood`: Reverse of the posterior probability.
    >- $P(y=k)$ is `Prior Probability`: Probability of an event that is calculated before considering the new information obtained.
    >- $P(X)$ is also known as `evidence` or normalization constant.

##  c. Examples of Bayes' Theorem

- **Example 1:** What is the probability that a card drawn from a deck of playing cards is a Queen (Q), given that it is a card of Spades (S)?

<font face = "Courier New" color=green size=4>$$P(y=Q \mid X) \hspace{0.5cm} = \hspace{0.5cm}  \frac{P(y=Q) \times P(X \mid y=Q)}{P(X)}$$</font>

<br><br>
$ \hspace{0.9cm} P(y=Q) = \large \frac{4}{52} , \hspace{0.9cm}\small P(X) = \large \frac{13}{52}, \hspace{0.9cm}\small P(X \mid y=Q) = \large \frac{1}{4}$

<br><br>

\begin{equation}
    P(y=Q \mid X) \hspace{0.5cm} = \hspace{0.5cm} \frac{P(y=Q) * P(X \mid y=Q)} {P(X)}\hspace{0.5cm}=\hspace{0.5cm} \frac{\frac{4}{52} * \frac{1}{4}}{\frac{13}{52}}\hspace{0.5cm} = \hspace{0.5cm}\frac{1}{13}
\end{equation}

<br><br>
- **Example 2:** In a school 60% of boys play football (F) and 36% of boys play cricket (C). The percentage of boys who play cricket given that they also play football is 40%. What is the percentage of those who play football given that they also play cricket?

$ \hspace{0.9cm} P(F) = 0.6 , \hspace{0.9cm} P(C) = 0.36 , \hspace{0.9cm} P(C \mid F) = 0.4, \hspace{0.9cm} P(F \mid C) = \hspace{0.2cm}? $

\begin{equation}
    P(F \mid C) \hspace{0.5cm} = \hspace{0.5cm} \frac{P(F) * P(C \mid F)}{P(C)} \hspace{0.5cm}
\end{equation}

\begin{equation}
    P(F \mid C) \hspace{0.5cm} = \hspace{0.5cm} \frac{0.4 * 0.6}{0.36} \hspace{0.5cm}= \hspace{0.5cm} \frac{4}{6} 
\end{equation}


**Example 3:** Consider a factory where three machines (M1, M2, M3) are manufacturing 20, 30 and 50 items respectively per day. Moreover, the machines M1, M2, and M3 are manufacturing  5%, 3% and 1% defective items out of the total items they produce. Find the probability that an item selected at random from a batch is manufactured by M3 given that it is defective? 
<br><br>
$\hspace{0.9cm} P(y=M1) = \large \frac{20}{100} \small= 0.2 , \hspace{2cm} P(y=M2) = \large \frac{30}{100} \small= 0.3 , \hspace{1.8cm} P(y=M3) = \large \frac{50}{100} \small= 0.5 $

$\hspace{0.9cm} P(X \mid y=M1) = \large \frac{5}{100} \small= 0.05 , \hspace{0.9cm} P(X\mid y=M2) = \large \frac{3}{100} \small= 0.03 , \hspace{0.9cm} P(X\mid y=M3) = \large \frac{1}{100} \small= 0.01 $
<br><br>

<br><br>
<font face = "Courier New" color=red size=4>
$$P(y=M3 \mid X) \hspace{0.5cm} = \hspace{0.5cm} \frac{P(y=M3) \times P(X \mid y=M3)}{P(X)} \hspace{8cm}$$
</font>

<br><br>
$$P(y=M3 \mid X) \hspace{0.5cm} = \hspace{0.5cm} \frac{P(y=M3) \times P(X \mid y=M3)}{P(X \cap M1)+P(X \cap M2)+ P(X \cap M3)}\hspace{3cm}.... (i)\hspace{2cm}$$<br><br>


<br>
We know that,
\begin{equation}
    P(A \mid B) = \frac{P(A \cap B)}{P(B)} \hspace{1cm} \implies \hspace{1cm}  P(A \cap B) =   P(A \mid B) * P(B) \hspace{9cm}
\end{equation}
<br>
By substituting in equation (i)

<br><br>
$$P(y=M3 \mid X) \hspace{0.5cm} = \hspace{0.5cm} \frac{P(y=M3) \times P(X \mid y=M3)}{P(X \mid y=M1)*P(y=M1)+P(X \mid y=M2)*P(y=M2)+ P(X \mid y=M3)*P(y=M3)}$$<br><br>
$$\hspace{5cm}=\hspace{0.5cm} \frac{0.5 * 0.01}{0.05*0.2 + 0.03*0.3 + 0.01*0.5} \hspace{0.5cm}\small =\hspace{0.5cm}0.2083 \hspace{10cm}$$

# 2. Overview of Naïve Bayes' Classifier

<h3 align="center"><div class="alert alert-success" style="margin: 20px">
The Naïve Bayes' (NB) is a probabilistic machine learning algorithm that use Bayes’ Theorem for supervised machine learning classification.
</h3>

- The Naïve Bayes' models are probabilistic classifiers, i.e., they not only assign a class label to a given sample, but they also provide an estimate of the probability that it belongs to that class. For example, a Naïve Bayes' model can predict that a given email has 80% chance of being a spam and 20% chance of being a ham.
    - `Deterministic classifiers` output a hard label for each sample, without providing probability estimates for the classes. Examples for such classifiers include K-nearest neighbors, decision trees, and SVMs.
    - `Probabilistic classifiers` output probability estimates for the k classes, and then assign a label to the sample based on these probabilities (typically the label of the class with the highest probability). Examples for such classifiers include Naïve Bayes' classifiers, logistic regression, and neural networks that use logistic/softmax in the output layer.
- The Naïve Bayes' classifier is called **Naïve** because it assumes that the input variables/features are independent of each other, which is rarely the case. In other words, changing the value of one feature, does not directly change the value of any of the other features. This assumption is naïve because it is almost never true.
- For example, if you like to predict a certain disease, based on the body weight and blood pressure, the method assumes that these two variables are independent, which means that a person's blood pressure does not depend on his/her weight.   
- Similarly, if you have temperature and humidity as input features, Naïve Bayes' assumes that temperature and humidity are independent of each other. That means changing the value of temperature does not directly change the value of humidity. However, in reality, we know that temperature and humidity are very much related to each other.
- Even though Naïve Bayes' is Naïve, it performs very well in applications, even when the features are not idependent of each other. Further more, compared to other ML algorithms, NB is fast, so it could be used for making predictions in real time.

# 3. Naïve Bayes' Classifier for Datasets with Discrete Input Features

<img align="right" width="270"  src="images/nb2a.png"  >

## a. Example 1 (Single Input Feature)
- Given the training dataset, build a Naïve Bayes' Classifier, so that given a new instance of fruit having `green` color, you should be able to predict the category of the fruit being apple or orange.
- The idea is to compute the two probabilities, that is the probability of the fruit being apple or orange given its color. Which ever fruit type has the highest probability wins.
- Mathematically, we need to compute and later compare the following two conditional probabilities, and whichever, probability is greater, we predict the fruit accordingly.

<font face = "Arial" color=magenta size=4>$\hspace{1cm} P(y=apple \mid X=green)\hspace{0.5cm}$<font face = "Arial" color=black size=4> and <font face = "Arial" color=blue size=4>$\hspace{0.5cm} P(y=orange \mid X=green)$</font>

<br><br>
- The Bayes' Theorem formula for predicting these probabilities is:

<font face = "Courier New" color=magenta size=4>$$     P(y=apple \mid X=green) \hspace{0.5cm} = \hspace{0.3cm}  \frac{P(y=apple) \times P(X=green \mid y=apple)}{P(X=green)}$$</font>

    
    
<font face = "Courier New" color=blue size=4>$$     P(y=orange \mid X=green) \hspace{0.5cm} = \hspace{0.3cm}  \frac{P(y=orange) \times P(X=green \mid y=orange)}{P(X=green)}$$</font>


**Step 1:** Create `contingency or frequency table` for each feature of training data (Using Panda's `crosstab()` method). Since we have only one feature, which is color, so. we only need to build on frequency table as shown below:

<img align="center" width="500"  src="images/nb1.png"  >


**Step 2:** Compute the `marginal` probabilities for each of the output label (apple, orange):<br>
$\hspace{2cm} P(y=apple) = \large\frac{54}{100}\hspace{1cm}$
$\hspace{2cm} P(y=orange) = \large\frac{46}{100}$

**Step 3:** Compute the `marginal` probabilities for the only input feature:<br>
$\hspace{2cm} P(X=red) = \large\frac{33}{100}\hspace{1cm}$
$\hspace{2cm} P(X=green) = \large\frac{27}{100}\hspace{1cm}$
$\hspace{2cm} P(X=orange) = \large\frac{40}{100}$

**Step 4:** Compute the `conditional` probabilities:<br>
$\hspace{2cm} P(X=green \mid y=apple) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{20}{54}$
$\hspace{4cm} P(X=green \mid y=orange) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{7}{46}$


**Step 5:** Substitute all the probabilities into the `Naïve Bayes' Formula`:<br>
$\hspace{2cm} P(y=apple \mid X=green) \hspace{0.8cm} = \hspace{0.3cm}  \large \frac{P(y=apple) \times P(X=green \mid y=apple)}{P(X=green)}\hspace{0.5cm} = \hspace{0.5cm} \frac{\frac{54}{100} * \frac{20}{54}}{\frac{27}{100}} \hspace{0.5cm}= \hspace{0.5cm} \frac{20}{27} $<br>


$\hspace{2cm} P(y=orange \mid X=green) \hspace{0.5cm} = \hspace{0.3cm}  \large \frac{P(y=orange) \times P(X=green \mid y=orange)}{P(X=green)}\hspace{0.5cm} = \hspace{0.5cm} \frac{\frac{46}{100} * \frac{7}{46}}{\frac{27}{100}} \hspace{0.5cm}= \hspace{0.5cm} \frac{7}{27} $<br>

<font face = "Arial" color=blue size=4>$\hspace{2cm}$ Since <br>  
<font face = "Arial" color=magenta size=4>$\hspace{4cm}P(y=apple \mid X=green) \hspace{0.1cm} \gt \hspace{0.1cm} P(y=orange \mid X=green),\hspace{0.3cm}$ <br><br>
<font face = "Arial" color=blue size=4>$\hspace{2cm}$ therefore, we classify the fruit given that its color is green is an apple
    
>- In our case, the output variable (y) has only two outcomes (apple or orange). So we will compute the probability of all these features being apple as well as the probability of all these features being orange. Whatever, probability is larger is our output label.

###  Simplified Formula of Naïve Bayes'
- Note that in step 5 above, the value of denominators of both formulas remain constant for a given input, i.e., $\large \frac{27}{100}$. 
- Since for Naïve Bayes' Classifier, we just need to compare these two probabilities, therefore, we can remove or ignore the term in the denominator. So for Naive Bayes' Classifier you can simplify the formula to 
    
<font face = "Courier New" color=green size=6>$$     P(y=k \mid X) \hspace{0.5cm} \propto \hspace{0.5cm}  P(y=k) \times P(X \mid y=k) \hspace{0.5cm} $$</font>

<img align="right" width="400"  src="images/nb1.png"  >

###  Zero Frequency Problem
- Suppose we are given a new instance of fruit having `red` color this time.
- In the frequency table, if you observe that out of 46 oranges you have zero orange having red color.
- Following the above described steps, we can use the Naïve Bayes' Classifier formula:
<br><br><br>
$$ P(y=apple \mid X=red) \hspace{0.8cm} = \hspace{0.3cm} P(y=apple) \times P(X=red \mid y=apple)\hspace{0.9cm} = \hspace{0.5cm} \small \frac{54}{100} * \frac{33}{54}\hspace{0.5cm}= \hspace{0.5cm} \frac{33}{100} \hspace{.5cm}=0.33\hspace{3cm}$$


$$ P(y=orange \mid X=red) \hspace{0.5cm} = \hspace{0.3cm} P(y=orange) \times P(X=red \mid y=orange)\hspace{0.5cm} = \hspace{0.5cm} \small\frac{46}{100} * \frac{0}{46} \hspace{0.5cm}= \hspace{0.5cm} 0 \hspace{4.5cm}$$

<font face = "Arial" color=blue size=4>$\hspace{2cm}$ Since <br>  
<font face = "Arial" color=magenta size=4>$\hspace{4cm}P(y=apple \mid X=green) \hspace{0.1cm} \gt \hspace{0.1cm} P(y=orange \mid X=green),\hspace{0.3cm}$ <br><br>
<font face = "Arial" color=blue size=4>$\hspace{2cm}$ therefore, we classify the fruit given that its color is red is an apple
    
> This seems to be working fine as there is only one input feature, i.e., color of the fruit. However, if we have many input features, and one of the input feature has a zero frequency, then the entire probability will become zero. This is called `Zero Frequency` problem, and it needs to be avoided. We can use Laplace smoothing to solve the problem of zero probability.

### Laplace Smoothing
- **Laplace Smoothing** is usually applied by adding one count to the numerator and adding number of possible values of features to the denominator. In this example number of possible values of the only input feature is 3 (red, green, orange). So the probability after Laplace smoothing is:
<br><br>
$$ P(y=apple \mid X=red) \hspace{0.8cm} = \hspace{0.3cm} P(y=apple) \times P(X=red \mid y=apple)\hspace{0.9cm} = \hspace{0.5cm} \small \frac{54}{100} * \frac{33+1}{54+3}\hspace{0.5cm}= \hspace{0.5cm} 0.3221 \hspace{3cm}$$


$$ P(y=orange \mid X=red) \hspace{0.5cm} = \hspace{0.3cm} P(y=orange) \times P(X=red \mid y=orange)\hspace{0.5cm} = \hspace{0.5cm} \small\frac{46}{100} * \frac{0+1}{46+3} \hspace{0.5cm}= \hspace{0.5cm} 0.00938\hspace{3cm}$$

<font face = "Arial" color=blue size=3>Even after Laplace Smoothing, we still classify the fruit having red color as apple, however, if we have multiple input features, the entire probability will not become zero <br>  

## b. Example 2 (Multiple Input Features)
<img align="center" width="400"  src="images/nbdata1.png"  > 

<font face = "Courier New" color=green size=6>$$ P(y=k \mid X_i) \hspace{0.5cm} \propto \hspace{0.5cm} P(y=k) \times P(X_i \mid y=k)\hspace{0.5cm} $$</font><br><br>
<font face = "Courier New" color=black size=5>$$ P(y=k \mid x_1,x_2,x_3,...x_m)  \hspace{0.5cm} \propto \hspace{0.5cm} P(y=k) * P(x_1 \mid y=k) * P(x_2 \mid y=k)  \cdots * P(x_m \mid y=k)$$</font>
<font face = "Courier New" color=black size=5>$$  P(y=k \mid x_1,x_2,x_3,...x_m)  \hspace{0.5cm} \propto \hspace{0.5cm} P(y=k) * \prod_{i = 1}^{m} P(x_i \mid y=k)\hspace{8.5cm}$$</font>
<font face = "Courier New" color=magenta size=5>$$     P(y=k \mid x_1,x_2,x_3,...x_n)  \hspace{0.5cm} = \hspace{0.5cm}\hat{y} \hspace{0.5cm} = \hspace{0.5cm}argmax_y P(y=k) * \prod_{i = 1}^{m} P(x_i \mid y=k)\hspace{6cm}$$</font>


>- **This is called a MAP (Maximum A-Posteriori) decision rule, since it chooses the hypothesis that maximizes the posterior probability.**

<img align="right" width="300"  src="images/nbdata1.png"  > 


- **Let us assume that on a particular day the outlook is `coludy`, temperature is `cool` and humidity is `normal`.** 


- Given the training dataset, build a Naïve Bayes' Classifier, so that given a new set of input features we should be able to predict if it will rain on that day?


- Mathematically, we need to compute and later compare the following two conditional probabilities, and whichever, probability is greater, we predict accordingly.    
    
<font face = "Courier New" color=green size=5>$P(y=yes \mid x_1=cloudy,x_2=cool,x_3=normal) \hspace{0.5cm}$</font>
<br><br>
<font face = "Courier New" color=green size=5>$P(y=no \mid x_1=cloudy,x_2=cool,x_3=normal) \hspace{0.5cm}$</font>


**Step 1:** Create contingency or frequency table for each feature of training data (Using Panda's `crosstab()` method). Since we have three features (outlook, temperature and humidity), so we need to construct three frequency tables as shown below:
<img align="center" width="1200"  src="images/freq-tablesa.png"  >

**Step 2:** Compute the probabilities for each of the output label (yes, no):<br>
$\hspace{2cm} P(y=yes) = \large\frac{6}{15}\hspace{1cm}$
$\hspace{2cm} P(y=no) = \large\frac{9}{15}$


**Step 3:** Compute the conditional probabilities:<br>
$\hspace{2cm} P(x_1=cloudy \mid y=yes) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{5}{6}$
$\hspace{2cm} P(x_2=cool \mid y=yes) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{4}{6}$
$\hspace{2cm} P(x_3=normal \mid y=yes) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{2}{6}$

$\hspace{2cm} P(x_1=cloudy \mid y=no) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{3}{9}$
$\hspace{2cm} P(x_2=cool \mid y=no) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{2}{9}$
$\hspace{2cm} P(x_3=normal \mid y=no) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{2}{9}$


<br><br>
**Step 4:** Substitute all the probabilities into the Naïve Bayes' Formula:<br>

<font face = "Courier New" color=green size=5>$P(y=yes \mid x_1=cloudy,x_2=cool,x_3=normal) \hspace{0.5cm}$</font>
<br><br>
$\hspace{3cm}=\hspace{0.5cm}P(y=yes)*P(x_1=cloudy|y=yes)*P(x_2=cool|y=yes)*P(x_3=normal|y=yes)$
<br><br>
$\hspace{3cm}=\hspace{0.5cm}\large \frac{6}{15}*\frac{5}{6}*\frac{4}{6}*\frac{2}{6}\hspace{0.5cm}\small =\hspace{0.5cm}\large \frac{240}{3240}\hspace{0.5cm}\small =\hspace{0.5cm}0.074\hspace{7.2cm}$


<br><br>
<font face = "Courier New" color=green size=5>$P(y=no \mid x_1=cloudy,x_2=cool,x_3=normal) \hspace{0.5cm}$</font>
<br><br>
$\hspace{3cm}=\hspace{0.5cm}P(y=no)*P(x_1=cloudy|y=no)*P(x_2=cool|y=no)*P(x_3=normal|y=no)$
<br><br>
$\hspace{3cm}=\hspace{0.5cm}\large \frac{9}{15}*\frac{3}{9}*\frac{2}{9}*\frac{2}{9}\hspace{0.5cm}\small =\hspace{0.5cm}\large \frac{108}{10935}\hspace{0.5cm}\small =\hspace{0.5cm}0.009\hspace{7.2cm}$

<br><br>
<font face = "Arial" color=blue size=4>$\hspace{1cm} Since, $<br>$  
$ <font face = "Courier New" color=green size=4>$\hspace{2cm}P(y=yes \mid x_1=cloudy,x_2=cool,x_3=normal) \hspace{0.5cm} \gt 
$</font> 
<font face = "Courier New" color=green size=4>$\hspace{0.5cm}P(y=no \mid x_1=cloudy,x_2=cool,x_3=normal) \hspace{0.5cm}$</font>    
    
    
<font face = "Arial" color=blue size=4>$\hspace{1cm}$ therefore, we predict that it will rain

## c. Example 3 (Multiple Input Features)

In [1]:
import pandas as pd
df = pd.read_csv('datasets/titanic2.csv')
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [2]:
df1 = pd.concat([df['pclass'], df['sex'], df['survived']], axis=1)
df1

,pclass,sex,survived
0,3,male,0
1,1,female,1
2,3,female,1
3,1,female,1
4,3,male,0
...,...,...,...
886,2,male,0
887,1,female,1
888,3,female,0
889,1,male,1


**Suppose we select a passenger at random from Titanic dataset, given that the passenger is female and belong to pclass 1, determine whether the passenger survived or not?**

**Create Contingency Tables** 

In [3]:
pd.crosstab(index=df1['sex'], columns=df1['survived'], margins=True)

survived,0,1,All
sex,,,
female,81,233,314
male,468,109,577
All,549,342,891


In [4]:
pd.crosstab(index=df1['pclass'], columns=df1['survived'], margins=True)

survived,0,1,All
pclass,,,
1,80,136,216
2,97,87,184
3,372,119,491
All,549,342,891


<img align="center" width="1000"  src="images/nb-ex2a.png"  >

<font face = "Courier New" color=green size=5>$P(y=survived \mid x_1=female,x_2=pclass1) \hspace{0.5cm}$</font>
<br><br>
$\hspace{3cm}=\hspace{0.5cm}P(y=survived)\hspace{0.3cm}*\hspace{0.3cm}P(x_1=female|y=survived)\hspace{0.3cm}*\hspace{0.3cm}P(x_2=pclass1|y=survived)$
<br><br>
$\hspace{3cm}=\hspace{0.5cm}\large \frac{342}{891}*\frac{233}{342}*\frac{136}{342}\hspace{0.5cm}\small =\hspace{0.5cm}0.1039\hspace{7.2cm}$


<br><br><br><br>
<font face = "Courier New" color=green size=5>$P(y=dead \mid x_1=female,x_2=pclass1) \hspace{0.5cm}$</font>
<br><br>
$\hspace{3cm}=\hspace{0.5cm}P(y=dead)\hspace{0.3cm}*\hspace{0.3cm}P(x_1=female|y=dead)\hspace{0.3cm}*\hspace{0.3cm}P(x_2=pclass1|y=dead)$
<br><br>
$\hspace{3cm}=\hspace{0.5cm}\large \frac{549}{891}*\frac{81}{549}*\frac{80}{549}\hspace{0.5cm}\small =\hspace{0.5cm}0.0132\hspace{7.2cm}$


<font face = "Arial" color=blue size=4>$\hspace{1cm} Since, $<br>$  
$ <font face = "Courier New" color=green size=4>$\hspace{2cm}P(y=survived \mid x_1=female,x_2=pclass1) \hspace{0.5cm} \gt 
$</font> 
<font face = "Courier New" color=green size=4>$\hspace{0.5cm}P(y=dead \mid x_1=female,x_2=pclass1) \hspace{0.5cm}$</font>    
    
    
<font face = "Arial" color=blue size=4>$\hspace{1cm}$ therefore, we predict that passenger will survive

## d. Example 4 (Text Data)
<img align="center" width="400"  src="images/nb-nlpa.png"  > 

- **Given a new sentence `I am playing cricket and happy`, Can you classify it as `Positive` or `Negative`.** 
<br><br><br>

**Step 1:** Construct the contingency table containing frequency count for all the words:
<img align="right" width="400"  src="images/nlp-tbl.png"  > 
<br><br><br><br><br><br>
<font face = "Courier New" color=green size=4>$P(y=+ \mid x_1=I,x_2=am, x_3=playing, x_4=cricket, x_5=happy) \hspace{0.5cm}$</font>
<br><br><br><br>
<font face = "Courier New" color=green size=4>$P(y=- \mid x_1=I,x_2=am, x_3=playing, x_4=cricket, x_5=happy) \hspace{0.5cm}$</font>


**Step 2:** Compute the probabilities for each of the output label (Positive, Negative):<br>
$\hspace{2cm} P(y=+) = \large \frac{1}{2}\hspace{1cm}$
$\hspace{2cm} P(y=-) = \large \frac{1}{2}$


**Step 3:** Compute the conditional probabilities:<br>
$P(x_1=I \mid y=+) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{3}{13}$
$\hspace{1cm} P(x_2=am \mid y=+) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{3}{13}$
$\hspace{1cm} P(x_3=playing \mid y=+) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{1}{13}$<br>
$P(x_4=cricket \mid y=+) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{1}{13}$
$\hspace{1cm} P(x_5=happy \mid y=+) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{2}{13}$

<br><br>

$P(x_1=I \mid y=-) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{3}{13}$
$\hspace{1cm} P(x_2=am \mid y=-) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{3}{13}$
$\hspace{1cm} P(x_3=playing \mid y=-) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{1}{13}$<br>
$P(x_4=cricket \mid y=-) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{1}{13}$
$\hspace{1cm} P(x_5=happy \mid y=-) \hspace{0.3cm} = \hspace{0.3cm} \large \frac{1}{13}$
<br><br>

**Step 4:** Substitute all the probabilities into the Naïve Bayes' Formula:<br>
<font face = "Courier New" color=green size=5>$P(y=+ \mid x_1=I,x_2=am, x_3=playing, x_4=cricket, x_5=happy) \hspace{0.5cm}$</font>
<br><br>
$\hspace{1cm}=\hspace{0.2cm}P(y=+)\hspace{0.1cm}*\hspace{0.1cm}P(x_1=I|+)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=am|y=+)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=playing|y=+)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=cricket|y=+)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=happy|y=+)$
<br><br>
$\hspace{1cm}=\hspace{0.5cm}\large \frac{1}{2}*\frac{3}{13}*\frac{3}{13}*\frac{1}{13}*\frac{1}{13}*\frac{2}{13} \small \hspace{0.5cm}=\hspace{0.5cm}0.000024$


<br><br><br>




<font face = "Courier New" color=green size=5>$P(y=- \mid x_1=I,x_2=am, x_3=playing, x_4=cricket, x_5=happy) \hspace{0.5cm}$</font>
<br><br>
$\hspace{1cm}=\hspace{0.2cm}P(y=-)\hspace{0.1cm}*\hspace{0.1cm}P(x_1=I|-)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=am|y=-)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=playing|y=-)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=cricket|y=-)\hspace{0.1cm}*\hspace{0.1cm}P(x_2=happy|y=-)$
<br><br>
$\hspace{1cm}=\hspace{0.5cm}\large \frac{1}{2}*\frac{3}{13}*\frac{3}{13}*\frac{1}{13}*\frac{1}{13}*\frac{1}{13} \small \hspace{0.5cm}=\hspace{0.5cm}0.0000121$


<br><br><br>
<font face = "Arial" color=blue size=4>$\hspace{1cm} Since, $<br>$  
$ <font face = "Courier New" color=green size=4>$\hspace{2cm}P(y=+ \mid x_1=I,x_2=am, \cdots x_5=happy) \hspace{0.5cm} \gt 
$</font> 
<font face = "Courier New" color=green size=4>$\hspace{0.5cm}P(y=- \mid x_1=I,x_2=am, \cdots x_5=happy) \hspace{0.5cm}$</font>    
    
    
<font face = "Arial" color=blue size=4>$\hspace{1cm}$ therefore, we classify that the given sentence belong to Positive class

# 4. Naïve Bayes' Classifier for Datasets with Continuous Input Features
## a. Example 1 (Single Input Feature)
- Consider this hypothetical data set showing the measures of the Prostate-Specific Antigen (PSA) concerntration from blood samples of fourteen persons, out of which seven are patients with prostate cancer while seven are healthy.
- Dataset has one input feature that is PSA level which is a continuous variable, while the output label has two possible values `Cancer` and `Healthy`, and the objective is to build a Naïve Bayes Classifier to predict if a given patient suffers from cancer or not given his/her PSA value.
- `100$ Question`: How to compute the probabilities when X is a continuous variable?
- `Answer`: Use  **Gaussian Naïve Bayes** model, which assumes that the continuous input feature variables follow a Gaussian (Normal) distribution.


<img align="center" width="650"  src="images/gnb1c.png"  >

>- **Suppose we like to predict if a person with `PSA level of 2.6` $\mu g/L $ has prostate cancer or not based on the given training data set of the two groups.**



**Calculate Sample Mean and Standard Deviation of Healthy group:**
- `Healthy Group:`

$ \hspace{2cm}\large \mu_{healthy} = \frac{\sum_{i=1}^n x_i}{n} = 1.5$<br>
<br><br>
$\hspace{2cm} \large \sigma_{healthy}  = \sqrt{\frac{\sum(x_i - \mu)^2}{n-1}}=0.61$<br>


- If we plug in the $\mu_{healthy}$ and $\sigma_{healthy}$ values for the healthy group with different $x_i$ values of PSA in the probability density function of Gaussian Distribution, we get different values which if plotted will give the following bell shaped curve for Healthy persons:

<font face = "Arial" color=magenta size=5> $$f(x) = \frac{1}{\sqrt{2\pi \sigma^2} }\hspace{0.4cm}\large e^{-\frac{(x-\mu)^2}{2\sigma^2}} $$ </font>


<img align="center" width="500"  src="images/gnb3.png"  >

**Calculate Sample Mean and Standard Deviation of all Groups:**
<img align="right" width="550"  src="images/gnb4a.png"  >
<br><br><br><br><br>
- `Healthy Group:`
$\hspace{1.8cm} \large \mu_{healthy} \small =1.5, \hspace{2cm} \large \sigma_{healthy} \small=0.61$
<br><br><br><br>
- `Cancer Group:`
$\hspace{2cm} \large \mu_{cancer} \small =2.8, \hspace{2cm} \large \sigma_{cancer} \small=0.82$

- Now we need to calculate $f(psa=2.6 \mid Healthy)$, and $f(psa=2.6 \mid Cancer)$ for which we will use the probability density function of Gaussian Distribution as shown below:

<img align="center" width="1200"  src="images/gnb6a.png"  >

> These values 0.13 and 0.47 are not the probabilities, rather are the values of the probability density function for the height of these curves where PSA=2.6

- Substitute these values in the Naïve Bayes' Formula:<br>

<font face = "Courier New" color=green size=4>$P(y=Healthy \mid psa=2.6) \hspace{0.5cm}=\hspace{0.5cm}P(y=Healthy)*P(psa=2.6|y=Healthy)\hspace{0.5cm}=\hspace{0.5cm}\large \frac{1}{2}\small *0.13\hspace{0.5cm} =\hspace{0.5cm}0.065$


<br><br>
<font face = "Courier New" color=orange size=4>$P(y=Cancer \mid psa=2.6) \hspace{0.5cm}=\hspace{0.5cm}P(y=Cancer)*P(psa=2.6|y=Cancer)\hspace{0.5cm}=\hspace{0.5cm}\large \frac{1}{2}\small *0.47\hspace{0.5cm} =\hspace{0.5cm}0.235$
    
    
<br><br><br>
<font face = "Arial" color=blue size=4>$\hspace{1cm} Since, $<br>$  
$ <font face = "Courier New" color=orange size=4>$\hspace{2cm}P(y=Cancer \mid psa=2.6) \hspace{0.5cm} \gt 
$</font> 
<font face = "Courier New" color=green size=4>$\hspace{0.5cm}P(y=Healthy \mid psa=2.6) \hspace{0.5cm}$</font>    
    
    
<font face = "Arial" color=blue size=4>$\hspace{1cm}$ therefore, we classify the person has Cancer.

<img align="right" width="300"  src="images/gnb7.png"  >

## b. Example 2 (Multiple Input Features)

- Consider this hypothetical dataset having two input features `PSA` level and `age`.
<br><br>

**Calculate Sample Mean and Standard Deviation of all Groups:**

- `Cancer Group:`<br>
$\hspace{2cm} \large \mu_{psa} \small =2.8, \hspace{2cm} \large \sigma_{psa} \small=0.82$<br>
$\hspace{2cm} \large \mu_{age} \small =67, \hspace{2cm} \large \sigma_{age} \small=6.45$

<br><br>
- `Healthy Group:`<br>
$\hspace{1.8cm} \large \mu_{psa} \small =1.5, \hspace{2cm} \large \sigma_{psa} \small=0.61$<br>
$\hspace{2cm} \large \mu_{age} \small =66, \hspace{2cm} \large \sigma_{age} \small=4.58$


<br><br><br>
>- **Suppose we like to predict if a person with `PSA level of 2.6` $\mu g/L $ and `age=70 years` has prostate cancer or not based on the given training data set of the two groups.**


**Calculate the conditional probabilities:**<br><br>
- `Cancer Group:`<br>
$f(psa=2.6 \mid Cancer) = \frac{1}{\sqrt{2\pi \sigma^2} }\hspace{0.4cm}\large e^{-\frac{(x-\mu)^2}{2 \sigma^2}} \hspace{.5cm}=\hspace{.5cm}\frac{1}{\sqrt{2\times 3.146 \times 0.82^2} }\hspace{0.4cm}\large e^{-\frac{(x-2.8)^2}{2\times 0.82^2}} \hspace{.5cm}=\hspace{.5cm} \small 0.47$<br>
$f(age=70 \mid Cancer) = \frac{1}{\sqrt{2\pi \sigma^2} }\hspace{0.4cm}\large e^{-\frac{(x-\mu)^2}{2 \sigma^2}} \hspace{.5cm}=\hspace{.5cm}\frac{1}{\sqrt{2\times 3.146 \times 6.45^2} }\hspace{0.4cm}\large e^{-\frac{(x-67)^2}{2\times 6.45^2}} \hspace{.5cm}=\hspace{.5cm}\small 0.055$

<br><br>
- `Healthy Group:`<br>
$f(psa=2.6 \mid Healthy) = \frac{1}{\sqrt{2\pi \sigma^2} }\hspace{0.4cm}\large e^{-\frac{(x-\mu)^2}{2 \sigma^2}} \hspace{.5cm}=\hspace{.5cm}\frac{1}{\sqrt{2\times 3.146 \times 0.61^2} }\hspace{0.4cm}\large e^{-\frac{(x-1.5)^2}{2\times 0.61^2}} \hspace{.5cm}=\hspace{.5cm} \small 0.13$<br>
$f(age=70 \mid Healthy) = \frac{1}{\sqrt{2\pi \sigma^2} }\hspace{0.4cm}\large e^{-\frac{(x-\mu)^2}{2 \sigma^2}} \hspace{.5cm}=\hspace{.5cm}\frac{1}{\sqrt{2\times 3.146 \times 4.58^2} }\hspace{0.4cm}\large e^{-\frac{(x-66)^2}{2\times 4.58^2}} \hspace{.5cm}=\hspace{.5cm}\small 0.059$

- Substitute all the probabilities into the Naïve Bayes' Formula:<br>
<font face = "Courier New" color=orange size=4>
$P(y=Cancer \mid psa=2.6, age=70)\hspace{0.5cm}=\hspace{0.5cm}P(y=Cancer)*P(psa=2.6|y=Cancer)*P(age=70|y=Cancer)$<br>
$\hspace{8.1cm}=\hspace{0.5cm}\large \frac{1}{2}\small *0.47*0.055\hspace{0.5cm} =\hspace{0.5cm}0.013$</font><br>

<br><br>
<font face = "Courier New" color=green size=4>
$P(y=Healthy \mid psa=2.6, age=70)\hspace{0.5cm}=\hspace{0.5cm}P(y=Healthy)*P(psa=2.6|y=Healthy)*P(age=70|y=Healthy)$<br>
$\hspace{8.1cm}=\hspace{0.5cm}\large \frac{1}{2}\small *0.13*0.059\hspace{0.5cm} =\hspace{0.5cm}0.004$</font><br>

    
<br><br><br>
<font face = "Arial" color=blue size=4>$\hspace{1cm} Since, $<br>$  
$ <font face = "Courier New" color=orange size=4>$\hspace{2cm}P(y=Cancer \mid psa=2.6, age=70) \hspace{0.5cm} \gt 
$</font> 
<font face = "Courier New" color=green size=4>$\hspace{0.5cm}P(y=Healthy \mid psa=2.6, age=70) \hspace{0.5cm}$</font>    
    
    
<font face = "Arial" color=blue size=4>$\hspace{1cm}$ therefore, we classify the person has Cancer.

# Tasks To Do

## Task 1:

- **Given the dataset, use Naïve Bayes' Classifier and predict if the  new instance having values A=0, B=1, C=0, belong to positive class or negative class.**

<img align="center" width="400"  src="images/nb-todo1.png"  >



## Task 2:

- **Given the dataset, use Gaussian Naïve Bayes' Classifier and predict the gender of a person having height 6 feet, weight 130 lbs, and foot size 8 inches.**


<img align="center" width="400"  src="images/nb-todo2.png"  >


## Task 3:

- **Given the dataset, use Naïve Bayes' Classifier and classify a new observation: Outlook = `Overcast`, Temp = `60`, Humidity=`62` , Wind= `False`, into either one of the two classes Yes or No.**

<img align="center" width="500"  src="images/nb-todo3.png"  >